### PR TITLE
feat(rpc): Add `admin_nodeInfo` to get the local node record

### LINKED ISSUE #850: Add an RPC method to get the local node record
### Describe the feature

We should add an RPC method named `admin_nodeInfo` that returns the ENR of the node. This depends on #849.

Additionally, `AdminApi` in the `reth-rpc` crate should be exported and constructable: currently it is private and impossible to construct.

### Additional context

To add the new RPC method, first extend the `AdminApiServer` trait here: https://github.com/paradigmxyz/reth/blob/main/crates/net/rpc-api/src/admin.rs

Then add the implementation for the new method here: https://github.com/paradigmxyz/reth/blob/main/crates/net/rpc/src/admin/mod.rs

The ENR should be fetched using the `NetworkHandle` in the `AdminApi` implementation.

### PR DESCRIPTION
**Pull request by Evalir**
Closes #850.

Adds the rpc method `admin_nodeInfo`, which returns the local node's ENR.

I'd like to ask about this part of the issue:

> Additionally, AdminApi in the reth-rpc crate should be exported and constructable: currently it is private and impossible to construct.

AdminApi seems to have been exported with #866 : https://github.com/paradigmxyz/reth/blob/main/crates/net/rpc/src/lib.rs#L19 but what does constructable mean? just providing a `new` method?

### CODE REVIEW

**Comment by onbjerg** [comment]
> but what does constructable mean? just providing a new method?

Yes, either that or making the fields `pub`, otherwise we cannot construct the struct. Not sure why none of the structs are constructable, cc @mattsse, maybe there is a reason

**Review by onbjerg — APPROVED** [review]
LGTM otherwise

File: crates/net/rpc-api/src/admin.rs [hunk 1]
```diff
@@ -33,4 +33,8 @@ pub trait AdminApi {
         item = String
     )]
     fn subscribe(&self);
+
+    /// Returns the ENR of the node.
```
Comment (mattsse) [inline crates/net/rpc-api/src/admin.rs:37]: please have a look at:
https://geth.ethereum.org/docs/interacting-with-geth/rpc/ns-admin

we should return the same object here

sowe we need:
1. NodeInfo struct in rpc-types
2. populate this in `node_info`

**Review by mattsse — CHANGES_REQUESTED** [review]
see `admin_nodeInfo`

fyi you can convert json into struct types via: https://app.quicktype.io/

**Comment by mattsse** [comment]
> > but what does constructable mean? just providing a new method?
> 
> Yes, either that or making the fields `pub`, otherwise we cannot construct the struct. Not sure why none of the structs are constructable, cc @mattsse, maybe there is a reason

yeh, let's just add a new function

File: crates/net/rpc-api/src/admin.rs (hunk 1 above)
Comment (Evalir) [reply crates/net/rpc-api/src/admin.rs:37]: ahh my bad! should've asked for more context before just jumping into the issue. Will do so! also, the quicktype utility is super neat!

**Comment by onbjerg** [comment]
FWIW the `admin_` namespace is not standardized and many nodes do it in different ways.

File: crates/net/rpc-types/src/admin.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    pub enode: String,
+    /// ID of the local node.
+    pub id: String,
+    /// IP of the local node.
+    pub ip: String,
+    /// Address exposed for listening for the local node.
+    #[serde(rename = "listenAddr")]
+    pub listen_addr: String,
+    /// Local node client name.
+    pub name: String,
```
Comment (Evalir) [inline crates/net/rpc-types/src/admin.rs:21]: Seems that to properly fill this property we'd need to pull:
- The version + (stable/unstable)
- OS
- Rust version

Any pointers on how we should do so for this situation, or will we roll our own response here?

File: crates/net/rpc-types/src/admin.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+pub struct Eth {
+    /// Total difficulty of the best chain.
+    pub difficulty: U256,
+    /// The hash of the genesis block.
+    pub genesis: String,
+    /// Hash of the latest block of the best chain.
+    pub head: String,
+    /// Network ID in base 10.
+    pub network: u64,
+}
```
Comment (Evalir) [inline crates/net/rpc-types/src/admin.rs:72]: Do we have a way of pulling these from this crate already? Probably related to the syncing parts of Reth

**Review by Evalir — COMMENTED** [review]
Created the corresponding types, some things left to do yet though:

**Review by mattsse — CHANGES_REQUESTED** [review]
great progress. 

A couple of suggestions and pointers.

File: crates/net/rpc-types/src/admin.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+impl NodeInfo {
+    /// Creates a new instance of `NodeInfo`.
+    pub fn new(enr: NodeRecord) -> NodeInfo {
+        NodeInfo {
+            enode: enr.to_string(),
+            id: enr.id.to_string(),
+            ip: enr.address.to_string(),
+            listen_addr: enr.tcp_addr().to_string(),
+            name: "Reth".to_owned(),
+            ports: Ports { discovery: enr.tcp_port.into(), listener: enr.tcp_port.into() },
```
[the remark this answers is not in the document]
Comment (gakonst) [reply crates/net/rpc-types/src/admin.rs:38]: Let's leave this as Reth and open an issue for making it OS-aware and versioned later?

File: crates/net/rpc-types/src/admin.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+pub struct Ports {
+    /// Port exposed for node discovery.
+    pub discovery: u16,
+    /// Port exposed for listening.
+    pub listener: u16,
+}
+
+/// Networking protocols being run by the node.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Protocols {
```
[the remark this answers is not in the document]
Comment (gakonst) [reply crates/net/rpc-types/src/admin.rs:57]: What would the key/value of the map be? Key probably string, value?

File: crates/net/rpc/src/admin.rs [hunk 6]
```diff
@@ -11,6 +13,13 @@ pub struct AdminApi {
     network: NetworkHandle,
 }
 
+impl AdminApi {
+    /// Creates a new instance of `AdminApi`.
+    pub fn new(network: NetworkHandle) -> AdminApi {
```
[the remark this answers is not in the document]
Comment (gakonst) [reply crates/net/rpc/src/admin.rs:18]: Which specific trait?

**Review by gakonst — COMMENTED** [review]
Good progress

File: crates/net/rpc-types/src/admin.rs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+impl NodeInfo {
+    /// Creates a new instance of `NodeInfo`.
+    pub fn new(enr: NodeRecord) -> NodeInfo {
+        NodeInfo {
+            enode: enr.to_string(),
+            id: enr.id,
+            ip: enr.address,
+            listen_addr: enr.tcp_addr(),
+            name: "Reth".to_owned(),
+            ports: Ports { discovery: enr.tcp_port.into(), listener: enr.tcp_port.into() },
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/net/rpc-types/src/admin.rs]: Done in https://github.com/paradigmxyz/reth/pull/880/commits/ebe85ce2ac4c305dfe3ed8a4ec9480b45b0932e8

File: crates/net/rpc-types/src/admin.rs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+/// Represents the `admin_nodeInfo` response, which can be queried for all the information
+/// known about the running node at the networking granularity.
+///
+/// Note: this format is not standardized.  Reth follows geth's format,
+/// see: https://geth.ethereum.org/docs/interacting-with-geth/rpc/ns-admin
+#[derive(Serialize, Deserialize, Debug)]
+pub struct NodeInfo {
+    /// Enode in URL format.
+    pub enode: String,
```
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/net/rpc-types/src/admin.rs]: Done in https://github.com/paradigmxyz/reth/pull/880/commits/ebe85ce2ac4c305dfe3ed8a4ec9480b45b0932e8

File: crates/net/rpc-types/src/admin.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/net/rpc-types/src/admin.rs]: @gakonst that sounds good for now! will open that issue. @mattsse gotcha—seems like something like `impl BlockProvider` is what we need, but I was trying to add `reth-provider` as a dep but seems like this triggers a cyclic dep issue with `reth-db`?

File: crates/net/rpc/src/admin.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Evalir) [reply crates/net/rpc/src/admin.rs:18]: Gotcha—I assume we wanna use the traits in `network-api`? if so, I have the same question as georgios—which traits, and, how are we gonna fill in the functionality that `NetworkHandle` has? I was actually gonna ask the same question in #683 as I'm assigned there too, so it's a good segway :D

File: crates/net/rpc-types/src/admin.rs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+impl NodeInfo {
+    /// Creates a new instance of `NodeInfo`.
+    pub fn new(enr: NodeRecord) -> NodeInfo {
+        let mut protocol_info = BTreeMap::from([(
+            "eth".into(),
+            ProtocolInfo::Eth(EthProtocolInfo {
+                difficulty: todo!(),
+                head: todo!(),
+                network: 0,
+                genesis: todo!(),
```
Comment (gakonst) [inline crates/net/rpc-types/src/admin.rs:41]: Can get that from the status message of the network component? I think we had a getter? @mattsse

Comment (mattsse) [reply crates/net/rpc-types/src/admin.rs:41]: yeh will open an issue to add a function for getting the status

**Review by mattsse — CHANGES_REQUESTED** [review]
just needs some lint fixes, then we can merge this and improve one we have a `network_status` function which should provide all of the missing info

**Comment by Evalir** [comment]
aight @mattsse made the lint fixes (it was the `todo!()`s tripping everything up)—should then be g2g pending ci!

**Review by gakonst — APPROVED** [review]
nice work - let's send it and add a way to keep the local node record up to date

**Pull request merged**